flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/xa/JdbcExactlyOnceSinkE2eTest.java

```diff
@@ -343,10 +360,15 @@ public void snapshotState(FunctionSnapshotContext context) {
         }
 
         private void sleep(Supplier<Boolean> condition) {
+            long start = System.currentTimeMillis();
             while (condition.get()
                     && running
                     && !Thread.currentThread().isInterrupted()
                     && haveActiveSources()) {
+                if (System.currentTimeMillis() - start > 10_000) {
+                    LOG.debug("Slept more than 10s", new Exception());
```

Comment:
Is this enough to understand why it hangs? I guess through the stacktrace we know for which condition it hangs...

Reply:
I added it just in case, I didn't see any tasks waiting in the latest failures. I think stacktrace + subtaskIdx + attemptId should be enough.